dsymbol/src/dsymbol/ufcs.d

```diff
@@ -26,43 +42,106 @@ enum string[string] INTEGER_PROMOTIONS = [
 
 enum MAX_RECURSION_DEPTH = 50;
 
+private DSymbol* deduceSymbolType(DSymbol* symbol)
+{
+    DSymbol* symbolType = symbol.type;
+    while (symbolType !is null && (symbolType.qualifier == SymbolQualifier.func
+            || symbolType.kind == CompletionKind.functionName
+            || symbolType.kind == CompletionKind.importSymbol
+            || symbolType.kind == CompletionKind.aliasName))
+    {
+        if (symbolType.type is null || symbolType.type is symbolType)
+        {
+            break;
+        }
+        //look at next type to deduce
+        symbolType = symbolType.type;
+    }
+    return symbolType;
+
+}
+
 // Check if beforeDotSymbol is null or void
-bool isInvalidForUFCSCompletion(const(DSymbol)* beforeDotSymbol)
+private bool isInvalidForUFCSCompletion(const(DSymbol)* beforeDotSymbol)
 {
     return beforeDotSymbol is null
         || beforeDotSymbol.name is getBuiltinTypeName(tok!"void")
         || (beforeDotSymbol.type !is null && beforeDotSymbol.type.name is getBuiltinTypeName(
                 tok!"void"));
 }
-/**
- * Get symbols suitable for UFCS.
- *
- * a symbol is suitable for UFCS if it satisfies the following:
- * $(UL
- *  $(LI is global or imported)
- *  $(LI is callable with $(D beforeDotSymbol) as it's first argument)
- * )
- *
- * Params:
- *     completionScope = current scope
- *     beforeDotSymbol = the symbol before the dot (implicit first argument to UFCS function)
- *     cursorPosition = current position
- * Returns:
- *     callable an array of symbols suitable for UFCS at $(D cursorPosition)
- */
-DSymbol*[] getSymbolsForUFCS(Scope* completionScope, const(DSymbol)* beforeDotSymbol, size_t cursorPosition)
+
+private TokenCursorResult getCursorToken(const(Token)[] tokens, size_t cursorPosition)
 {
-    if (beforeDotSymbol.isInvalidForUFCSCompletion)
+    auto sortedTokens = assumeSorted(tokens);
+    auto sortedBeforeTokens = sortedTokens.lowerBound(cursorPosition);
+
+    TokenCursorResult tokenCursorResult;
+
+    if (sortedBeforeTokens.empty) {
+        return tokenCursorResult;
+    }
+
+    if (sortedBeforeTokens.length >= 2 
+        && sortedBeforeTokens[$ - 1].type is tok!"."
+        && sortedBeforeTokens[$ - 2].type is tok!"identifier")
     {
-        return null;
+        // Check if it's UFCS dot completion
+        tokenCursorResult.completionContext = UFCSCompletionContext.DotCompletion;
+        tokenCursorResult.symbolIdentifierName = istring(sortedBeforeTokens[$ - 2].text);
+        return tokenCursorResult;
+    }
+    else
+    {
+        // Check if it's UFCS paren completion
+        size_t index = goBackToOpenParen(sortedBeforeTokens);
+
+        if (index == size_t.max)
+        {
+            return tokenCursorResult;
+        }
+
+        auto slicedAtParen = sortedBeforeTokens[0 .. index];
+        if (slicedAtParen.length >= 4
+            && slicedAtParen[$ - 4].type is tok!"identifier"
+            && slicedAtParen[$ - 3].type is tok!"."
```

Comment:
ditto for this may be `foo(x).ufcsBar(y)`

Reply:
function chaining isn't yet supported, which is why I needed the return type implementation earlier